File: Array/Exercises.java
import java.util.*;

public class Exercises {
	
	/**
	* Array Intro - Find Pivot Index
	*/
	public static int pivotIndex(int[] nums) {
        
        // Add all elements and assign as right
        int right = 0;
        for (int item: nums) {
        	right += item;
        }
        // Iterate elements add to left and subtract from right
        // until left equals right and return index
        int left = 0;

        for (int i = 0; i < nums.length; i++) {
        	right -= nums[i];
        	if (left == right) {
        		return i;
        	}
        	left += nums[i];
        }
        return -1;

    }

    /**
    * Array Intro - Largest Number At Least Twice of Others
    */
    public static int dominantIndex(int[] nums) {

        // Find the biggest element
        int biggest = 0;
        for(int item: nums) {
            if (item > biggest) {
                biggest = item;
            }
        }
        int result = -1;
        // Iterate elements
        for (int i = 0; i < nums.length; i++) {
            // Return index of biggest element
            if (nums[i] == biggest) {
                result = i;
                continue;
            } 
             // Return -1 if twice of element is greater than biggest element
            if ((nums[i] * 2) > biggest) {
                result = -1;
                return result;
            }
        }
        return result;
    }

    /**
    * Array Intro - Plus One
    */
    public static int[] plusOne(int[] digits) {
        int[] results = new int[digits.length + 1];
        int dIndex = digits.length - 1;
        int rIndex = dIndex + 1;
        int sum = digits[dIndex] + 1;
        int carry = sum / 10;
        results[rIndex] = sum % 10;
        dIndex--;
        rIndex--;
        while (dIndex >=0) {
            sum = digits[dIndex] + carry;
            carry = sum / 10;
            results[rIndex] = sum % 10;
            dIndex--;
            rIndex--;
        }
        if (carry > 0) {
            results[rIndex] = 1;
            return results;
        }
        return Arrays.copyOfRange(results, rIndex + 1, results.length); 
    }

    public static void main(String[] args) {
    	// // Test LeetCode : Array Intro - Find Pivot Index
    	// int[] nums = {1, 7, 3, 6, 5, 6};
    	// System.out.println(pivotIndex(nums));

        // Test Array Intro - Largest Number At Least Twice of Others
        // int[] nums1 = {3, 6, 1, 0};
        // int[] nums2 = {1, 2, 4, 3};
        // System.out.println(dominantIndex(nums1));
        // System.out.println(dominantIndex(nums2));

        // Test Array Intro - Plus One
        int[] nums1 = {3, 6, 1, 0};
        int[] nums2 = {9, 9, 9, 9};
        System.out.println(plusOne(nums2));
    }
}

File: Array/Exercises.py
class Solution:

    def pivotIndex(self, nums):
        """
        Array Intro - Find Pivot Index
        """
        # Compute the prefix sums for each index
        right = sum(nums)
        left = 0
        for index, num in enumerate(nums):
            right -= num
            if left == right:
                return index
            left += num
        return -1

    def dominantIndex(self, nums):
        """
        Array Intro - Largest Number At Least Twice of Others
        :type nums: List[int]
        :rtype: int
        """
        biggest = max(nums)
        for index, num in enumerate(nums):
            if num == biggest:
                result = index
            elif num * 2 > biggest:
                return -1
        return result

    def plusOne(self, digits):
        """
        Array Intro - Plus One
        :type digits: List[int]
        :rtype: List[int]
        """
        result = []
        index = len(digits) - 1
        total = digits[index] + 1
        carry = total // 10
        result.insert(0, total % 10)
        while index > 0:
            index -= 1
            total = digits[index] + carry
            carry = total // 10
            result.insert(0, total % 10)
        if carry == 1:
            result.insert(0, 1)
        return result

    

if __name__ == '__main__':

    # # Test Array Intro - Find Pivot Index
    # print(Solution().pivotIndex([1, 3, 7, 6, 5, 6]))

    # # Test Array Intro - Largest Number At Least Twice of Others
    # print(Solution().dominantIndex([3, 6, 1, 0]))

    # Test Array Intro - Plus One
    # print(Solution().plusOne([9, 9, 9]))
    # print(Solution().plusOne([9, 9, 8]))

